### ISSUE TITLE
h$gcInterval default value does not fit some applications

### ISSUE THREAD

**Issue by ryantrinkle** [ghcjs/ghcjs#550]
I've found that in many cases, performing a GC every second is too often.  In most of my applications, I set the GC interval to 1 minute instead, and this seems to work pretty well.

Ideally, the system would have a heuristic that could determine when to run the GC.  However, as a stopgap measure, an adjustment of the default value might help.

**Comment by 3noch** [comment]
I'm curious what your typical peak memory usage is with such a long GC interval?

**Comment by ryantrinkle** [comment]
Our apps often have resident sets of a couple hundred megabytes.  I can get more precise data if necessary.

**Comment by ryantrinkle** [comment]
Yeah, on a decent machine it's a small stutter - I haven't clocked it recently, but I think it's around half a second.

**Comment by dmjio** [comment]
+1 for adjustment of GC interval at the application level (as opposed to during ghcjs build)

**Comment by luite** [comment]
I think it's already possible to change the interval at runtime. The default value can be adjusted with a link-time option.

But it's rather undocumented, and this, plus a bunch of other parameters that affect various aspects of the RTS are currently really just a mishmash of JS variables. There is no clear indication of reasonable value ranges.

The true long-term solution would be to eliminate noticeable pauses from the GC pass altogether, perhaps by making it incremental. New JS features may help in the future.

Shorter term, I think it would be easy to make the RTS self-tuning to some extent. The RTS would reduce GC frequency if the fraction of total time spent in the GC is higher than some threshold. I think a relatively simple algorithm would allow the GC to automatically get out of the way for the vast majority of applications, without the need for explicitly tweaking parameters.

**Comment by luite** [comment]
But I also think there should be a better way to tweak the parameters for those applications that still need it.

I think GHC has a bunch of FFI calls into C functions for this, but I don't think there's a general way to tweak runtime RTS parameters, or even a general way to discover which parameters can be adjusted at all.

What can we do here? Suggestions welcome!

**Comment by dmjio** [comment]
@luite, unsure but a few things come to mind.

1) Hooks in `Setup.hs` (can these affect options during linking?)
2) Command line options to cabal that affect linking, `cabal configure --option foo bar` (something like this)
3) (hackiest) new module in `ghcjs-base` where RTS options can be set when `main` gets run.

**Comment by chessai** [comment]
Where is the change of the default at link-time currently documented? 

edit: in an archived reddit thread - 
https://www.reddit.com/r/haskell/comments/7tiacx/how_do_i_set_longer_gc_interval_for_ghcjs/
cc @andrewthad @goolord
